sync-core/src/main/java/com/cloudant/http/CookieFilter.java

```diff
@@ -0,0 +1,188 @@
+                //don't resend request, failed to get cookie
+                if(cookie != null) {
+                    context.replayRequest = true;
+                } else {
+                    context.replayRequest = false;
+                }
+            }
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, "Failed to get response code from request",e);
+        }
+        return context;
+
+    }
+
+     private String getCookie(URL url){
+        try {
+            URL sessionURL = new URL(String.format("%s://%s:%d/_session",
+                    url.getProtocol(),
+                    url.getHost(),
+                    url.getPort()));
+
+            HttpConnection conn = Http.POST(sessionURL, "application/json");
+            conn.setRequestBody(sessionRequestBody);
+            HttpURLConnection connection = conn.execute().getConnection();
+            String cookieHeader = connection.getHeaderField("Set-Cookie");
+            int responseCode = connection.getResponseCode();
+
+            if(responseCode == 401){
+                shouldAttemptCookieRequest  = false;
+                logger.severe("Credentials are incorrect, cookie authentication will not be" +
+                        " attempted again");
+            } else if (responseCode / 100 == 5){
+                logger.log(Level.SEVERE,
+                        "Failed to get cookie from server, response code %s, cookie auth",
+                        responseCode);
+            } else if(responseCode / 100 == 2) {
+
+                if(responseLooksOkay(connection.getInputStream())) {
+                    return cookieHeader.substring(0, cookieHeader.indexOf(";"));
+                } else {
+                    return null;
+                }
+
+            } else {
+                // catch any other response code
+                logger.log(Level.SEVERE,
+                        "Failed to get cookie from server, response code %s, " +
+                                "cookie authentication will not be attempted again",
+                        responseCode);
+                shouldAttemptCookieRequest = false;
+            }
+
+        } catch (MalformedURLException e) {
+            logger.log(Level.SEVERE,"Failed to create URL for _session endpoint",e);
+        } catch (UnsupportedEncodingException e) {
+            logger.log(Level.SEVERE, "Failed to encode cookieRequest body", e);
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, "Failed to read cookie response header", e);
+        }
+        return null;
+    }
+
+    private boolean responseLooksOkay(InputStream responseStream){
+        //check the response bod
+        JSONHelper jsonHelper = new JSONHelper();
+        Map<String,Object> jsonResponse = jsonHelper.fromJson(new InputStreamReader(responseStream));
+        boolean responseLooksOkay = true;
+
+        responseLooksOkay = jsonResponse.containsKey("ok") && responseLooksOkay;
```

Comment:
This method looks messy and it's hard to understand what you're looking for in the JSON.

Why don't you let our JSON library do the mapping for you, by creating a class to capture the response, like here? https://github.com/cloudant/sync-android/blob/master/sync-core/src/main/java/com/cloudant/sync/replication/CouchClientWrapper.java#L88